### PR TITLE
 Add a new FdtNorFalshQemuLib and enable it in ArmVirtPkg

### PR DESCRIPTION
**Pull request by kilaterlee**
# Description

Add a new FdtNorFalshQemuLib and enable it in ArmVirt platforms

* Patch1: Added a new library called FdtNorFlashQemuLib in OvmfPkg/Library which is non-hardcode dependency.
* Patch2: Enable the new library in ArmVirtQemu.dsc and ArmVirtQemuKernel.dsc.

BZ: https://bugzilla.tianocore.org/show_bug.cgi?id=4770

### Step 1
Build the two platforms firmware using ArmVirtQemu.dsc and ArmVirtQemuKernel.dsc, and then create the pflash files using following command:
```
cat QEMU_EFI.fd  /dev/zero | head -c 64m > ./QEMU_EFI-pflash.raw
cat QEMU_VARS.fd  /dev/zero | head -c 64m > ./vars-template-pflash.raw
qemu-img convert -f raw -O qcow2 -o cluster_size=4096 -S 4096 QEMU_EFI-pflash.raw  QEMU_EFI-pflash.qcow2
qemu-img convert -f raw -O qcow2 -o cluster_size=4096 -S 4096 vars-template-pflash.raw vars-template-pflash.qcow2
```
### Step 2
Verification the pflash working:
#### ArmVirtQemu:
Run the QEMU ARM virt machine using the following command:
```
qemu-system-aarch64 \
    -blockdev '{"driver":"file","filename":"/usr/share/edk2/aarch64/QEMU_EFI-pflash.qcow2","node-name":"libvirt-pflash0-storage","auto-read-only":true,"discard":"unmap"}' \
    -blockdev '{"node-name":"libvirt-pflash0-format","read-only":true,"driver":"qcow2","file":"libvirt-pflash0-storage"}' \
    -blockdev '{"driver":"file","filename":"/usr/share/edk2/aarch64/vars-template-pflash.qcow2","node-name":"libvirt-pflash1-storage","auto-read-only":true,"discard":"unmap"}' \
    -blockdev '{"node-name":"libvirt-pflash1-format","read-only":false,"driver":"qcow2","file":"libvirt-pflash1-storage"}' \
    -machine virt,pflash0=libvirt-pflash0-format,pflash1=libvirt-pflash1-format,acpi=on \
    -cpu cortex-a57 \
    -m size=2097152k \
    -serial stdio \
    -net none \
    -device ramfb \
    -device nec-usb-xhci \
    -device usb-mouse \
    -device usb-kbd
```

#### ArmVirtQemuKernel:
Run the QEMU kernel ARM virt machine using the following command:
```
qemu-system-aarch64 \
    -blockdev '{"driver":"file","filename":"/usr/share/edk2/aarch64/vars-template-pflash.raw","node-name":"libvirt-pflash1-storage","auto-read-only":true,"discard":"unmap"}' \
    -blockdev '{"node-name":"libvirt-pflash1-format","read-only":false,"driver":"raw","file":"libvirt-pflash1-storage"}' \
    -machine virt,usb=off,dump-guest-core=off,gic-version=3,pflash1=libvirt-pflash1-format \
    -cpu cortex-a57 \
    -m 4096 \
    -smp 1,sockets=1,cores=1,threads=1 \
    -no-user-config \
    -nodefaults \
    -device virtio-gpu-pci \
    -kernel /usr/share/edk2/aarch64/QEMU_EFI-pflash.raw \
    -serial stdio \
    -device nec-usb-xhci \
    -device usb-mouse \
    -device usb-kbd  \
    -hda /usr/share/edk2/aarch64/openEuler-22.03-LTS-SP3-aarch64.qcow2 \
    -monitor tcp::3333,server,nowait
```

### Step 3
* After the virt-machines starts, enter "F2" to enter the setup UI, try to change the boot order or some ther variables stored in the flash, then enter "F10" to save the changes and reboot it.
* After restarting, enter "F2" to enter the setup UI and check whether the changes before the restart operation have been saved.

Using the above three steps, both platforms will works fine

## Integration Instructions

N/A

### CODE REVIEW

File: OvmfPkg/Library/FdtNorFlashQemuLib/FdtNorFlashQemuLib.c [hunk 1]
```diff
@@ -0,0 +1,165 @@
+/** @file
+
+  Copyright (c) 2014-2018, Linaro Ltd. All rights reserved.<BR>
+  Copyright (c) 2024 Loongson Technology Corporation Limited. All rights reserved.<BR>
```
Comment (ardbiesheuvel) [inline OvmfPkg/Library/FdtNorFlashQemuLib/FdtNorFlashQemuLib.c:4]: Which part of this file is original code? If none, drop this copyright claim.

**Review by ardbiesheuvel — COMMENTED** [review]
Why is this signed off by Xianglai Li <lixianglai@loongson.cn>?

File: ArmVirtPkg/ArmVirtQemu.dsc [hunk 2]
```diff
@@ -66,7 +66,7 @@
   QemuLoadImageLib|OvmfPkg/Library/GenericQemuLoadImageLib/GenericQemuLoadImageLib.inf
 
   TimerLib|ArmPkg/Library/ArmArchTimerLib/ArmArchTimerLib.inf
-  VirtNorFlashPlatformLib|ArmVirtPkg/Library/NorFlashQemuLib/NorFlashQemuLib.inf
+  VirtNorFlashPlatformLib|OvmfPkg/Library/FdtNorFlashQemuLib/FdtNorFlashQemuLib.inf
```
Comment (ardbiesheuvel) [inline ArmVirtPkg/ArmVirtQemu.dsc:69]: We can remove the old library entirely, no?

File: OvmfPkg/Library/FdtNorFlashQemuLib/FdtNorFlashQemuLib.c (hunk 1 above)
Comment (kilaterlee) [reply OvmfPkg/Library/FdtNorFlashQemuLib/FdtNorFlashQemuLib.c:4]: In fact, most of the code about frame is original, only new logic is added between lines 111 to 136, and the logic of lines 100 and 101 are adjusted.

File: ArmVirtPkg/ArmVirtQemu.dsc (hunk 2 above)
Comment (kilaterlee) [reply ArmVirtPkg/ArmVirtQemu.dsc:69]: Yes, the old version can be removed, should I add a patch to remove it?

**Comment by kilaterlee** [comment]
> Why is this signed off by Xianglai Li [lixianglai@loongson.cn](mailto:lixianglai@loongson.cn)?

Because he contributed code, and I change to "Co-authored-by" now.

**Review by kraxel — APPROVED** [review]

File: ArmVirtPkg/ArmVirtQemu.dsc (hunk 2 above)
Comment (ardbiesheuvel) [reply ArmVirtPkg/ArmVirtQemu.dsc:69]: yes please send a follow-up patch to remove it.

**Review by ardbiesheuvel — APPROVED** [review]

File: ArmVirtPkg/ArmVirtQemu.dsc (hunk 2 above)
Comment (kilaterlee) [reply ArmVirtPkg/ArmVirtQemu.dsc:69]: OK，I will send it today.

**Comment by ardbiesheuvel** [comment]
@mdkinney this PR was updated by the author while the 'push' label was set and the CI was running (and taking very long!)

IIUC this means the version that gets merged may differ from the version that i reviewed? Could we fix that please?

**Comment by kilaterlee** [comment]
> @mdkinney this PR was updated by the author while the 'push' label was set and the CI was running (and taking very long!)
> 
> IIUC this means the version that gets merged may differ from the version that i reviewed? Could we fix that please?

While this PR CI was running and the 'push' label was set, I noticed that this PR was not up to date, the bot had merged 18 patches since this PR CI started, so I rebased this PR and commit again. Am I doing something wrong?

**Comment by ardbiesheuvel** [comment]
> > @mdkinney this PR was updated by the author while the 'push' label was set and the CI was running (and taking very long!)
> > IIUC this means the version that gets merged may differ from the version that i reviewed? Could we fix that please?
> 
> While this PR CI was running and the 'push' label was set, I noticed that this PR was not up to date, the bot had merged 18 patches since this PR CI started, so I rebased this PR and commit again. Am I doing something wrong?

No, this is not your fault. The infrastructure should prevent you from modifying the PR once it has been reviewed and the 'push' label has been set by the maintainer. This is especially relevant given that completing the CI checks can take hours (if they complete at all) so the window for potential abuse can be rather large.

**Comment by kilaterlee** [comment]
> > > @mdkinney this PR was updated by the author while the 'push' label was set and the CI was running (and taking very long!)
> > > IIUC this means the version that gets merged may differ from the version that i reviewed? Could we fix that please?
> > 
> > 
> > While this PR CI was running and the 'push' label was set, I noticed that this PR was not up to date, the bot had merged 18 patches since this PR CI started, so I rebased this PR and commit again. Am I doing something wrong?
> 
> No, this is not your fault. The infrastructure should prevent you from modifying the PR once it has been reviewed and the 'push' label has been set by the maintainer. This is especially relevant given that completing the CI checks can take hours (if they complete at all) so the window for potential abuse can be rather large.

OK, I will update this PR when the CI checks down.
One more question, I noticed that this PR was added a new merge type patch by the robot before I rebase and submit again, I guess it's like the gerrit method: it adds merge message patch when the commit is not the latest, I can't stand it, very afraid of this happening, so how can we avoid this from happening?

**Pull request merged**